libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -271,539 +398,578 @@ toUMapPair (UMap !m1 !m2) =
   , "pointers" .= m2
   ]
 
--- | It is worthwhile stating the invariant that holds on a Unified Map
---   The 'ptrmap' and the 'ptrT' field of the 'tripmap' are inverses.
+instance Crypto c => EncCBOR (UMap c) where
+  encCBOR UMap {..} =
+    encodeListLen 2 <> encodeMap encCBOR encCBOR umElem <> encodeMap encCBOR encCBOR umPtr
+
+instance Crypto c => DecShareCBOR (UMap c) where
+  type Share (UMap c) = (Interns (Credential 'Staking c), Interns (KeyHash 'StakePool c))
+  decSharePlusCBOR =
+    StateT
+      ( \(a, b) ->
+          decodeRecordNamed "UMap" (const 2) $ do
+            umElem <- decodeMap (interns a <$> decCBOR) (decShareCBOR b)
+            let a' = internsFromMap umElem <> a
+            umPtr <- decodeMap decCBOR (interns a' <$> decCBOR)
+            pure (UMap {..}, (a', b))
+      )
+
+-- | It is worthwhile stating the invariant that holds on a Unified Map.
+-- The `umPtr` and the `ptrT` field of the `umElem` are inverses.
 umInvariant :: Credential 'Staking c -> Ptr -> UMap c -> Bool
-umInvariant stake ptr (UMap tripmap ptrmap) = forwards && backwards
+umInvariant cred ptr UMap {..} = forwards && backwards
```

Comment:
I'd suggest using `NamedFieldPuns` throughout this module, rather than `RecordWildCards`. Not much overhead, but it will make it a bit easier for the future reader.
```suggestion
umInvariant cred ptr UMap {umElem, umPtr} = forwards && backwards
```